src/inventory_ui.cpp

```diff
@@ -3088,15 +3145,44 @@ item_location inventory_pick_selector::execute()
 {
     shared_ptr_fast<ui_adaptor> ui = create_or_get_ui_adaptor();
     debug_print_timer( tp_start );
+    item *startDragItem = nullptr;
+    bool dragActive = false;
     while( true ) {
         ui_manager::redraw();
         const inventory_input input = get_input();
 
         if( input.entry != nullptr ) {
-            if( highlight( input.entry->any_item() ) ) {
-                ui_manager::redraw();
+            if( input.action == "CLICK_AND_DRAG" ) {
+                if( input.entry->any_item() != item_location::nowhere ) {
+                    dragActive = true;
+                    startDragItem = input.entry->locations.front().get_item();
+                }
+            } else if( input.action == "MOUSE_MOVE" ) {
+                if( !dragActive && highlight( input.entry->any_item() ) ) {
+                    ui_manager::redraw();
+                }
+            } else if( input.action == "SELECT" ) {
+                dragActive = false;
+                item *startDragItemCpy = startDragItem;
+                startDragItem = nullptr;
+                item *endDragItem = nullptr;
+                if( input.entry->any_item() != item_location::nowhere ) {
+                    endDragItem = input.entry->locations.front().get_item();
+                    if( startDragItemCpy != nullptr && endDragItem != nullptr && startDragItemCpy != endDragItem ) {
+                        if( drag_drop_item( startDragItemCpy, endDragItem ) ) {
+                            clear_items();
+                            add_character_items( u );
```

Comment:
You're assuming here that the pick selector was initialized with the character's inventory. Quick search shows that it's also used for reading ebooks, looking inside containers, and picking nearby CBMs for autodoc.

Atm I don't have a good solution to suggest. Typically we solve this by assigning a menu return activity.

Reply:
maybe we can add some sort of flag to enable drag-drop so that it only works when the selector is populated with the character's inventory.

Reply:
I added a boolean in inventory_pick_selector for enabling drag-drop support, only set to true for the main 'i' inventory screen.